Code/Framework/AzCore/AzCore/std/hash_table.h

```diff
@@ -165,32 +166,43 @@ namespace AZStd
 
-                    typename list_type::iterator newIter, iter(cur);
                     size_type numValues = 1;
-                    for (++iter; iter != last && table->m_keyEqual(Traits::key_from_value(*cur), Traits::key_from_value(*iter)); ++iter, ++numValues)
+                    // Get the number of same consecutive elements in the table with same key,
+                    // this allows range insertion of elements at once
+                    for (++curEnd; curEnd != last && table->m_keyEqual(valueKey, Traits::key_from_value(*curEnd)); ++curEnd, ++numValues)
                     {
                     }
-                    ;
 
-                    const typename HashTable::key_type& valueKey = Traits::key_from_value(*cur);
                     size_type newBucketIndex = table->bucket_from_hash(table->m_hasher(valueKey));
+
+                    // newBucket.first holds the total number of elements in the bucket
+                    // newBucket.second contains the pointer to the first element in the bucket
                     vector_value_type& newBucket = newBuckets[newBucketIndex];
                     size_type numElements = newBucket.first;
-                    newIter = newBucket.second;
+                    insertIter = newBucket.second;
+
+                    // If we don't have elements in the bucket yet, transfer the elements directly
                     if (numElements == 0)
                     {
-                        newList.splice(newList.begin(), m_list, cur, iter);
+                        newList.splice(newList.begin(), m_list, cur, curEnd);
                         newBucket.second = newList.begin();
                     }
                     else
                     {
-                        if (!table->find_insert_position(valueKey, table->m_keyEqual, newIter, numElements, integral_constant<bool, Traits::has_multi_elements>()))
+                        // Since there are elements already in the bucket, update `insertIter` to where the elements will need to be inserted.
+                        if (!table->find_insert_position(valueKey, table->m_keyEqual, insertIter, numElements, integral_constant<bool, Traits::has_multi_elements>()))
                         {
-                            continue;
+                            // An element was found but we don't allow for duplicate elements in this table.
+                            // This happens when there was an insertion of two elements that are equal but have different hashes,
+                            // which is undefined behavior for a hash table: ISO C++ N4713, section 23.14.15 - 5.3
+                            AZ_Assert(false, "Found a duplicate element when rehashing. "
+                                "Review the hashing function for type '%s' and make sure two equal elements always have the same hash", typeid(typename Traits::value_type).name());
```

Comment:
I wanted to include the type that didn't have the proper hashing function in the Assert to be more user-friendly.
Is there a better way to do this? If not, we can remove the type name

Reply:
Well you could use `AzTypeInfo.h` for getting information about the type name but that creates a dependency on the our TypeInfo system I would rather not be there as it makes it hard to decouple AZStd from AzCore.

I am OK with using the `typeid` keyword, but if you do you must provide the `<typeinfo>` include at the top of hash_table.h otherwise the calls to the typeid could fail in other .dll/.so/.dylib files due to those not bringing in info due to the missing include.

We have seen this issue before with the EditorPythonBindings Gem missing the include for typeinfo and causing undefined behavior.